deepspeed/runtime/zero/stage3.py

```diff
@@ -812,12 +771,40 @@ def __init__(self,
 
         self.reduce_bucket_size = int(reduce_bucket_size)
 
-        self.reduction_event = torch.cuda.Event(enable_timing=False, blocking=False)
+        # IPG
+        if contiguous_gradients:
+            self.__ipg_bucket_flat_buffer: Tensor = torch.empty(
+                int(reduce_bucket_size),
+                dtype=self.dtype,
+                device=torch.cuda.current_device())
+
+        self.__param_id_to_grad_partition: Dict[int, Tensor] = {}
+
+        all_params = list(itertools.chain.from_iterable(self.fp16_groups))
+
+        grad_partitions_flat_buffer: Tensor = torch.zeros(
+            sum(p.ds_tensor.ds_numel for p in all_params),
+            dtype=self.dtype,
+            device=self.device,
+            pin_memory=self.offload_optimizer_pin_memory)
+
+        offset = 0
+        for param in all_params:
+            self.__param_id_to_grad_partition[
+                param.ds_id] = grad_partitions_flat_buffer.narrow(
+                    0,
+                    offset,
+                    param.ds_tensor.numel())
+            offset += param.ds_tensor.numel()
+
+        self.__params_in_ipg_bucket: List[Parameter] = []
+        self.is_gradient_accumulation_boundary: bool = True
 
-        self.reduction_stream = torch.cuda.Stream(
-        ) if self.overlap_comm else torch.cuda.current_stream()
-        self.callback_queued = False
-        self.copy_grad_stream = torch.cuda.Stream()
+        self.__param_reduce_events: Deque[Event] = collections.deque()
+        self.__max_param_reduce_events: int = 2
```

Comment:
Does it make sense to make this a json configurable value?

Reply:
i wanted to avoid putting them in the public API since they won't be necessary anymore after PyTorch starts using stream ordered `cudaMallocAsync` and `cudaFreeAsync` in their caching memory allocator. 

Maybe we could have it read from env variable or something so savvy users can mess with it, but knowing that DeepSpeed team has the discretion to remove this at any time?

Reply:
but you will have users still using old pytorch for at least a year or even longer. Deepspeed currently still support pt-1.2, is it not?

it shouldn't be a problem to make things configurable and indicate in the docs that a param x is experimental or a subject to change - as long as the disclaimer is added from the beginning - backward compat doesn't need to be maintained and no "social" agreements will be broken.

Reply:
@jfc4050, I think you can add a TODO to make this configurable.

Reply:
sure that works for me

Reply:
added TODOs.